### ISSUE TITLE
Support sharing and viewing file-based "Tilt Snapshots"

### ISSUE THREAD

**Issue by andymartin-sch** [tilt-dev/tilt#2513]
The "Snapshot" functionality from https://docs.tilt.dev/snapshots.html relies on a central server (cloud.tilt.dev) to host all user snapshots.

Some users of Tilt may want to use this feature but are blocked by concerns about sensitive data leaking out.

EDIT: the idea in the next sentence has been superseded by other ones mentioned in the comment thread below. ~It would be nice if there was a self-hosted option that allowed users to (a) run their own snapshot server and (b) configure their Tilt processes to always upload snapshots to this server instead of cloud.tilt.dev.~

**Comment by andymartin-sch** [comment]
I've thought a little more about this.

A potentially better short-term approach to accomplish the underlying goal would be to:
1. add an API method like `disable_snapshots()` but called something like `force_local_snapshots()` and have it cause Tilt to write the snapshot payload to the local filesystem instead of uploading it to cloud.tilt.dev
2. add a new `tilt snapshot` command (like `tilt up`, `tilt doctor`, etc) that takes a local file as its sole argument. it would launch the Tilt server in the same way as `tilt up` but instead of starting services, it just visualizes the data from the snapshot file.

 @jazzdan / @nicks : This could also be applied in the Jenkins/CI user story we discussed last week (also mentioned in https://github.com/windmilleng/tilt/issues/988) in that CI tests that launch Tilt internally could produce this snapshot file as a build artifact. This file could be used for debugging build failures in a more user-friendly way than the raw logs from `tilt up --hud=false --watch=false`.

I've renamed this issue so that the title is more focused on the underlying goal of enabling users to use snapshots without worrying about secrets leaking into the cloud.

**Comment by andymartin-sch** [comment]
(this idea was briefly discussed with @jazzdan and @nicks yesterday)

If what's described in the previous comment gets implemented, an interesting extension would be to have the payload that Tilt writes to disk actually be _a self-contained HTML file_ that contains (a) enough of the Tilt UI's HTML/CSS/JS for basic functionality (just clicking on services and scrolling through logs) and (b) the JSON payload from before.

This would allow the HTML file to be downloaded from the build server (or sent around via email/chat, attached to JIRA tickets, etc) and users would just need to double click the file to open it up in a browser. Using a self-contained file in this way would also eliminate issues from mismatches in `tilt` versions across the producer & consumers of the snapshot files by letting consumers no longer need to have a local `tilt` binary present at all.

This would not have to break the `tilt snapshot $FILENAME` functionality from the previous comment, as that could be updated to just shell out to open the file in the user's web browser and quit when it detects the file is HTML and not a simple JSON.

**Comment by andymartin-sch** [comment]
just to put it in writing after the meeting today: @nicks (& other Tilt folks) please feel free to edit this issue however you all see fit (or close it and open a new one) if you think moving forward with either a `tilt snapshot $filename` mode or having Tilt output a full HTML "snapshot" file is a possibility 👍

**Comment by victorwuky** [comment]
@andymartin-sch : I've created https://github.com/windmilleng/tilt/issues/3067 specifically as one possibility to address the problem. Feel free to comment there. And definitely do 👍it as a signal that drives our prioritization of features. Thanks!

**Comment by victorwuky** [comment]
Closing this since we have https://github.com/tilt-dev/tilt/issues/3067 shipped with private snapshots now.
